### PR TITLE
add a next interceptor handler on DioInterceptor for missing mocks

### PR DESCRIPTION
**Pull request by sebastianbuechler**
This PR is a proposal for issue https://github.com/lomsa-dev/http-mock-adapter/issues/136.

This package has two ways of mocking for Dio:
- DioInterceptor
- DioAdapter

It is my understanding that selective mocking only works for the first one as the latter one completely exchanges the HttpClientAdapter of Dio.

My proposed solution is to add a boolean failOnMissingMock, which is by default true for backwards compatibility. If set to false we do not throw an exception anymore when a mock is not matched, but instead send back a Future.value(null). This way we can detect the mismatch in the DioInterceptor and let Dio call the next interceptor.

Also, I added an option to output logs which are by default turned off and use the common logger package: https://pub.dev/packages/logger

@LukaGiorgadze what do you think?

### CODE REVIEW

**Comment by LukaGiorgadze** [comment]
Thank you @sebastianbuechler for your contribution, our community appreciates it!
What about logging, It's really nice, but let's make it optional and users decide whether to use it or the standard one.

**Comment by sebastianbuechler** [comment]
@LukaGiorgadze So a solution that just allows to inject a logger into the `DioAdapter`/`DioInterceptor` you mean and not having one already in pubspec.yaml?

**Comment by LukaGiorgadze** [comment]
> @LukaGiorgadze So a solution that just allows to inject a logger into the `DioAdapter`/`DioInterceptor` you mean and not having one already in pubspec.yaml?

I think it't fine unless we are not too much dependent on that logger. So I'll approve PR.

**Review by LukaGiorgadze — APPROVED** [review]

**Comment by sebastianbuechler** [comment]
> > @LukaGiorgadze So a solution that just allows to inject a logger into the `DioAdapter`/`DioInterceptor` you mean and not having one already in pubspec.yaml?
> 
> I think it't fine unless we are not too much dependent on that logger. So I'll approve PR.

I was thinking about using the https://github.com/Milad-Akarie/pretty_dio_logger so that everybody can decide for themselves, but I saw that we have here some usecases (like "route matched"), that would not been covered with this approach. 

By the way, I'm not allowed to merge the PRs that are approved and open. Can you merge them? Also, do you want to make a new release afterward?

**Comment by LukaGiorgadze** [comment]
@sebastianbuechler I promoted you as a maintainer, now you should be able to push changes too.
Yeah, time to release, You should be able to do this as well, but I'll do it if you want.
By the way, I guess we should downgrade the collection before we release 0.6.x version.

**Comment by sebastianbuechler** [comment]
@LukaGiorgadze Thanks! I just did a rebase in order to have a clean state for merging. Do you want to downgrade the collection package via PR?

**Comment by LukaGiorgadze** [comment]
> @LukaGiorgadze Thanks! I just did a rebase in order to have a clean state for merging. Do you want to downgrade the collection package via PR?
Thanks, this is a proper way to update PR's before we merge.
Yep, please.

**Pull request merged**
